Add log_prob_elemwise feature

lucianopaz opened the pull request:
This closes #157.

### Core idea

I simply copied the ordinary `SamplingExecutor.collect_log_prob` to return a dictionary of `{"rv_name": elem_log_prob_value}` instead of reducing the `dist.log_prob(...)` with `tf.reduce_sum`.

Using this we can use `pm.evaluate_model(model())` to get a `SamplingState` which can then be used to `collect_log_prob_elemwise`.

### Ease of use

I think that it should be easy for an end-user to be able to get the output of `collect_log_prob_elemwise` without having to jump through so many hoops (evaluating the model, getting a `SamplingState`, finally calling the `SamplingState.collect_log_prob_elemwise`) so I added `model_log_prob_elemwise` which does all of that behind the scenes.

### Discussion

- `model_log_prob_elemwise` isn't really needed. It's just there to help end-users. Should it be moved to another part of pymc4?
- Is the interface I wrote for `model_log_prob_elemwise` OK?
- When we decide on `model_log_prob_elemwise`'s fate, I'll write it's docstring
- It would be great to test this with some multivariate distribution, where the `event_shape` isn't `()`.

kyleabeauchamp commented:
Thanks, will play with this PR soon.

kyleabeauchamp commented:
I'm having reasonable success with this approach.  For various reasons, I ended up needing to do something like the following:

```
    for i in range(n_steps):
        tf_trace = pm4.inference.sampling.sample(model_instantiated, ...)
        lp = pm4.model_log_prob_elemwise(model_instantiated)
        # Accumulate arrays etc
```

Eventually I'll try to add this as a deterministic inside the model, once that feature stabilizes.

kyleabeauchamp commented:
Actually, I guess if I pass `values=trace[0]` to the `model_log_prob_elemwise` function, is it going to run the calculation for each step in the MCMC trace?  I guess that avoids the above for loop that I mentioned, if I'm current in my interpretation of what it's intending to do.

lucianopaz commented:
Adding as a deterministic might be a problem. I'll ping @ferrine to hear his input, but within the model function's flow, the distribution instance is never available. The model function only sees the samples that come out of the distribution, i.e. the values.
To be able to see the distribution, we would need the model to somehow have a reference to the `SamplingState`.

kyleabeauchamp commented:
I think I'm fine without deterministic at this point.

On Wed, Oct 23, 2019, 11:40 PM Luciano Paz <notifications@github.com> wrote:

> Adding as a deterministic might be a problem. I'll ping @ferrine
> <https://github.com/ferrine> to hear his input, but within the model
> function's flow, the distribution instance is never available. The model
> function only sees the samples that come out of the distribution, i.e. the
> values.
> To be able to see the distribution, we would need the model to somehow
> have a reference to the SamplingState.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/pymc4/pull/158?email_source=notifications&email_token=AAOP7K5TLGZQH6KLUVN264TQQE7KRA5CNFSM4JCHHSA2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECD4ODY#issuecomment-545769231>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOP7KY24FMMCGPOEHI5YBDQQE7KRANCNFSM4JCHHSAQ>
> .
>

lucianopaz commented:
About your second point, I think that yes, you should be able to pass `values=trace[0]` to `model_log_prob_elemwise`. I haven't tried it out yet, though.

lucianopaz commented:
I rebased this and **removed** the high level `model_log_prob_elemenwise` API. In the end, this didn't look important to me, as we can wrap the distribution's `log_prob` in a `Deterministic` in the following way:

```python
@pm.model
def model():
    a = yield pm.Normal("a", 0, 1)
    b = yield pm.HalfNormal("b", 1)
    c_dist = pm.Normal("c", loc=a, scale=b, observed=np.random.randn(100))
    c = yield c_dist
    log_prob = yield pm.Deterministic("log_prob", c_dist._distribution.log_prob(c))
```

The key to this method is that we don't yield from the `pm.Distribution` immediately, so **we have access to the underlying `tfp.Distribution`**. This would also allow us to use the batch and event shape manipulations that `tfp` provides, such as `Independent`, `BatchReshape` and others.

On the other hand, I added a function to get an iterable over all the distribution and potential's unreduced `log_prob` values. This could prove useful if we want to add some way to allow users to manually handle batching to improve runtimes.

ColCarroll commented on pymc4/flow/executor.py:
```diff
@@ -143,12 +143,17 @@ def __init__(
+    def collect_log_prob(self):
+        return sum(map(tf.reduce_sum, self.collect_log_prob_elemwise()))
+
+    def collect_unreduced_log_prob(self):
```
this looks like a `ValueError` waiting to happen, unless all the elements are the same shape, right?

lucianopaz replied:
Yes

We are doing automatic batching in all three big sampling functions: `sample_prior_predictive`, `sample` and `sample_posterior_predictive`. To be able to get the auto batching to work in some cases brings an over-head of extra `evaluate_model*` calls. Furthermore, the auto batching will have the overhead of always creating `SamplingState`s and other intermediate entities for every batch element. So in the end, I think that if someone wants to get their hands dirty to manually ensure that their model is well vectorized and batched, then they should have the means to do so.

I agree that this function is dangerous. It could be a `ValueError` waiting to happen, or a silently well broadcasting thing that leads to mysterious errors down the way, but it provides the necessary interface for a user to manually vectorize and batch their model.

lucianopaz replied:
It's very hard to get the model to be well vectorized. If there are reductions, or tensor products, or even simple sums, the `batch_shape`s can lead to horribly wrong `log_prob` tensors. The automatic batching is super user friendly and awesome, but someone may ask for a more low level control of what goes on

ColCarroll replied:
i'm fine leaving this -- was just wondering if I was right or wrong about it being dangerous :D

ColCarroll commented on tests/test_executor.py:
```diff
@@ -509,6 +561,21 @@ def model():
+    log_prob = state.collect_log_prob()
+    assert len(log_prob_elemwise) == len(expected_rv_shapes)
+    assert all(rv in log_prob_elemwise for rv in expected_rv_shapes)
+    print({k: v.shape for k, v in log_prob_elemwise.items()})
```
can remove prints

lucianopaz replied:
Good catch
